#167729793 Users should book a return Trip

davidfrank96 opened the pull request:
#### What does this PR do?
** A User will be able to book a return Trip 
#### Description of Task to be completed?
- created a return Trip controller
- Created a Trips table
- created a validation Handler for handling errors
- created a date validation function
- created not empty function to validate the req.body
- created an Authorization controller to validate the user token
#### How should this be manually tested?
** npm pull from the origin Branch from staging
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[Finishes #167729793] 
#### Screenshots (if appropriate)
![ReturnTrip](https://user-images.githubusercontent.com/43038045/64151761-eb43c900-cddf-11e9-98a0-ca61de0d17d0.JPG)

#### Questions:
N/A

okabamac committed:
Initial commit

okabamac committed:
Initialize controllers

okabamac committed:
Initialize controllers

davidfrank96 committed:
working progress

okabamac committed:
Refactor validation to use camelcase

okabamac committed:
Refactor to use camel case

okabamac committed:
models, seed and migration setup

davidfrank96 committed:
feat(Return) Implement exaclty once
- created a return Trip controller
- Created a Trips table
- created a validation Handler for handling errors
- created a date validation function
- created notEmpty fuction to validate the req.body
- created an Authorization controller to validate the user token
[Finishes #167729793]

davidfrank96 committed:
feat(ResolveConflicts) Implement exaclty once
-deleted my authorization file
- added a test file
[Finishes #167729793]

chibytez commented on src/controllers/returnTrip.js:
```diff
@@ -0,0 +1,70 @@
+            if (!duration) {
+                return res.status(400).json({
+                    status: 'error',
+                    message: 'Departure date and Return date cant be the same'
```
Can you please fix your can't

chibytez commented on src/controllers/returnTrip.js:
```diff
@@ -0,0 +1,70 @@
+            const tripsResult = await Trips.create({
+                currentOfficeLocation,
+                tripId: uuidv4(),
+                user_id: id,
```
We are using camel case not snake case

chibytez commented on src/controllers/returnTrip.js:
```diff
@@ -0,0 +1,70 @@
+                return res.status(201).json({
+                    status: 'success',
+                    data: {
+                        user_id: id,
```
Deto

chibytez commented on src/controllers/returnTrip.js:
```diff
@@ -0,0 +1,70 @@
+                        destination,
+                        current_office_location: currentOfficeLocation,
+                        departure_date: departureDateUTC,
+                        return_date: returnDateUTC,
```
deto

chibytez commented on src/database/migrations/20190901024921-create-Trips.js:
```diff
@@ -0,0 +1,55 @@
+      type: Sequelize.STRING,
+      unique: true
+    },
+    user_id: {
```
Can we please change this to camel case

chibytez commented on src/test/returntrip.test.js:
```diff
@@ -0,0 +1,50 @@
+                requestStatus: 'pending',
+                destination: '',
+            })
+            .end((err, res) => {
```
Can we also take care of the err

davidfrank96 committed:
feat(Feedbacks)
- made some changes to my controller
- validated my request.body
- clean up some redundant code
[Finishes #167729793]

chibytez reviewed (changes requested):
Can we add a screenshot to show that the test is passing

chibytez commented on src/controllers/returnTrip.js:
```diff
@@ -0,0 +1,69 @@
+            if (!duration) {
+                return res.status(400).json({
+                    status: 'error',
+                    message: 'Departure and Return date can"t be the same'
```
can we fix it to can't

sebaztyn commented on src/models/Trip.js:
```diff
@@ -0,0 +1,51 @@
+            type: DataTypes.INTEGER,
+            required: true
+        },
+        departureDate: {
```
Based on @Stylll 's recommendation, this should be a date type

sebaztyn commented on src/models/Trip.js:
```diff
@@ -0,0 +1,51 @@
+            type: DataTypes.STRING,
+            required: true
+        },
+        returnDate: {
```
This should have a date type too

sebaztyn commented on src/models/Trip.js:
```diff
@@ -0,0 +1,51 @@
+            type: DataTypes.STRING,
+            required: true
+        },
+        destination: {
```
We are advised to make this an array

chibytez commented on src/routes/api/index.js:
```diff
@@ -11,6 +12,7 @@ usersRouter.use('/v1/users', passwordRouter);
 usersRouter.get('/', (req, res) => res.status(200).send('Welcome to  Shadowcat API'));
 
 usersRouter.use('/v1/users', passwordRouter);
+usersRouter.use('/v1/trips', returntrip);
```
I think it should be /v1/users/trips

sebaztyn replied:
All trip requests for oneway, return and multi-city should use just one route based on the advice of @Stylll

chibytez commented on src/routes/api/returntrip.js:
```diff
@@ -0,0 +1,12 @@
+
+const router = express.Router();
+
+router.post('/request', Auth.authenticate, returntripvalidate.returnTrip, ValidationHandler.validate,  ReturnTrip.returnTripRequest );
```
I think your validation should be the first route

chibytez commented on src/utils/ValidationHandler.js:
```diff
@@ -0,0 +1,46 @@
+
+            const mappedErrors = errors.mapped();
+
+            let results = Object.keys(mappedErrors).map(function (key) {
```
Can we use arrow function here

chibytez commented on src/utils/helper/notEmpty.js:
```diff
@@ -0,0 +1,19 @@
+ * @param {string} msg
+ * @return {(error|bool)} returns error or true
+ */
+function notEmpty(value, msg) {
```
deto

sebaztyn commented on src/test/returntrip.test.js:
```diff
@@ -0,0 +1,56 @@
+                reason: 'Holiday',
+                tripType: 'Return-trip',
+                requestStatus: 'pending',
+                destination: 'Lagos',
```
Destination should be array of integers and current location should be an integer representing the current office location

okabamac committed:
Refactor validation to cover for different trip type

davidfrank96 committed:
feat(bug)
- trying to fix a bug
[Finishes #167729793]

davidfrank96 committed:
feat(TripHandler)
- Added a Trip Handlermiddleware
[Finishes#167729793]

okabamac committed:
Initial commit

okabamac committed:
Set up controller for one-way trip request
Set up validations and test

okabamac committed:
Resolve merge conflict

okabamac committed:
Resolve merge conflict

okabamac committed:
ft(one-way-request):
- Allow users to request a one way trip
- Add validations and test to endpoint
- Refactor the multi-city validation
[Delivers #167729792]

okabamac committed:
ft(one-way-request):
- Allow users to request a one way trip
- Add validations and test to endpoint
- Refactor the multi-city validation
[Delivers #167729792]

okabamac committed:
--reuse-message=HEAD

davidfrank96 committed:
feat(returnTrip)
- changed the controller
[Finishes#167729793]

okabamac committed:
--reuse-message=HEAD

okabamac committed:
--reuse-message=HEAD

davidfrank96 committed:
feat(returnTrip)
- merging staging branch
[Finishes#167729793]

davidfrank96 committed:
Merge remote-tracking branch 'origin/ft-one-way-trip-request-167729792' into users-book-return-Trip-167729793

davidfrank96 committed:
feat(returnTrip)
-progress
[Finishes#167729793]

davidfrank96 committed:
feat(ReturnTrip)
-fixed the locations bug
[Finishes#167729793]

davidfrank96 committed:
feat(returnTrip)
-wrote test for this feat
- added a seed:test file to package.json file
[Finishes#167729793]

davidfrank96 committed:
feat(returnTrip)
-still working on the test
[Finishes#167729793]

davidfrank96 committed:
feat(ReturnTrip)
-removed Trip models
-refactored to es6
[Finishes#167729793]

davidfrank96 committed:
feat(ReturnTrip)
- fixed a NotEmpty function
[Finishes#167729793]

davidfrank96 committed:
feat(ReturnTrip)
-test
[Finishes#167729793]
